graph/src/components/store/entity_cache.rs

```diff
@@ -203,18 +202,94 @@ impl EntityCache {
             entity_type: EntityType::new(base_type.to_string()),
-            entity_field: field.name.clone().to_snake_case().into(),
+            entity_field: field.name.clone().into(),
             value: eref.entity_id.clone(),
             causality_region: eref.causality_region,
             id_is_bytes: id_is_bytes,
         };
 
-        let entities = self.store.get_derived(&query)?;
-        entities.iter().for_each(|(key, e)| {
-            self.current.insert(key.clone(), Some(e.clone()));
-        });
-        let entities: Vec<Entity> = entities.values().cloned().collect();
-        Ok(entities)
+        let mut entity_map = self.store.get_derived(&query)?;
+
+        for (key, entity) in entity_map.iter() {
+            // Only insert to the cache if it's not already there
+            // This is to avoid overwriting updates
+            if !self.current.contains_key(&key) {
+                self.current.insert(key.clone(), Some(entity.clone()));
+            }
+        }
+
+        // Apply updates from `updates` and `handler_updates` directly to entities in `entity_map` that match the query
+        for (key, entity) in entity_map.iter_mut() {
+            let mut entity_cow = Some(Cow::Borrowed(entity));
+
+            if let Some(op) = self.updates.get(key).cloned() {
+                op.apply_to(&mut entity_cow)
+                    .map_err(|e| key.unknown_attribute(e))?;
+            }
+
+            if let Some(op) = self.handler_updates.get(key).cloned() {
+                op.apply_to(&mut entity_cow)
+                    .map_err(|e| key.unknown_attribute(e))?;
+            }
+
+            if let Some(updated_entity) = entity_cow {
+                *entity = updated_entity.into_owned();
+            }
+        }
+
+        // A helper function to apply an update and return the resulting entity if it matches the query
+        fn apply_update(
+            op: &EntityOp,
+            query: &DerivedEntityQuery,
+            key: &EntityKey,
+        ) -> Result<Option<Entity>, anyhow::Error> {
+            match op {
+                EntityOp::Update(entity) | EntityOp::Overwrite(entity)
+                    if query.matches(key, entity) =>
+                {
+                    Ok(Some(entity.clone()))
+                }
+                EntityOp::Remove => Ok(None),
+                _ => Ok(None),
+            }
+        }
+
+        // Iterate over self.updates to find entities that:
+        // - Aren't already present in the entity_map
+        // - Match the query
+        // If these conditions are met:
+        // - Check if there's an update for the same entity in handler_updates and apply it.
+        // - Add the entity to entity_map.
+        for (key, op) in self.updates.iter() {
+            if !entity_map.contains_key(key) {
+                if let Some(entity) = apply_update(op, &query, key)? {
+                    if let Some(handler_op) = self.handler_updates.get(key) {
```

Comment:
If an entity is in both `self.updates` and `self.handler_updates`, that means it was created and then modified. In that case, don't we have to apply the `op` from `self.handler_updates` to the entity from `self.updates` just as we do in the case for entities that already exist in the store? I think that's where the name `apply_update` is misleading.

Reply:
`apply_update` oh thats right, exactly that bad naming that i did confused me as well. Just modified the name, comment and addresses the update from `handler_op` in this case